### PR TITLE
PSDP: Woodgate Algorithm

### PR DESCRIPTION
**Pull request by banrovegrie**
The basic algorithm has been implemented. Reformatting the code and testing are left. Putting in this PR so that we can discuss this tomorrow.

### CODE REVIEW

**Comment by banrovegrie** [comment]
@FanwangM let's run tests only after I have reformatted the code.

**Comment by banrovegrie** [comment]
I tested the algorithm for square matrices and it appears to work. 3 doubts remain which I have posted in the discussion forum.

**Comment by banrovegrie** [comment]
@FanwangM could you help me out on where tox is failing?

File: procrustes/woodgate.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        D = find_gradient(E=E, LE=LE_pos, G=G)
+
+        func = lambda w: f(E - w * D)
+        w_min = minimize(func, 1, bounds=((1e-9, None),)).x[0]
+        E = E - w_min * D
+        i += 1
+
+    print(f"Woodgate's algorithm took {i} iterations.")
+    print(f"Error = {np.linalg.norm(F - E.T @ E @ G)}.")
+    print(f"Required P = {E.T @ E}")
```
Comment (FanwangM) [inline procrustes/woodgate.py:259]: We don't need this information in the final version and it's safe to comment them out.

**Review by FanwangM — CHANGES_REQUESTED** [review]

File: procrustes/woodgate.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    return E.T @ E
```
Comment (FanwangM) [inline procrustes/woodgate.py:260]: We will return a `ProcrustesResult` object in the end.

**Comment by FanwangM** [comment]
Can you help finalize this algorithm as we discussed including the formatting style?
Please feel free to tag me if you have questions. Thank you. @banrovegrie

File: procrustes/woodgate.py (hunk 1 above)
Comment (banrovegrie) [reply procrustes/woodgate.py]: removed in 3f938e10c7d2069dbcced45d25a57b500bf60de1

File: procrustes/woodgate.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        - np.trace(arr.T @ arr @ Q)
+    )
+
+    # Main part of the algorithm.
+    i = 0
+    n = F.shape[0]
+    E = np.random.rand(n, n)
+
+    while True:
+        LE = L(E)
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: it has been renamed and is a lambda function at line 286

File: procrustes/woodgate.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    arr : np.ndarray
+        The array :math:`A` such that :math:`v(A') = Pv(A)`.
+
+    Returns
+    -------
+    np.ndarray
+        The permutation matrix.
+    """
+    k = 0
+    x, _ = arr.shape
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: fixed in 3f938e10c7d2069dbcced45d25a57b500bf60de1

File: procrustes/woodgate.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+#
+# --
+"""Positive semidefinite Procrustes Module."""
+
+import numpy as np
+import scipy.linalg as lin
+from scipy.optimize import minimize
+# from procrustes.utils import ProcrustesResult
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: There is an issue with using ProcrastesResult, I will elaborate below,

**Comment by banrovegrie** [comment]
@FanwangM The algorithm has exact precision as expected. After implementing scaling and choosing $E_0 = I_n$, I obtained the exact results mentioned in the paper's examples.

Our algorithm also converges equally as fast.

File: procrustes/woodgate.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+__all__ = ["woodgate"]
+
+
+def permutation_matrix(arr: np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py:34]: Oh this method is not related to the permutation Procrustes problem. It just deals with finding a matrix such that $v(E^T') = Pv(E)$.

Here, $v$ represents flattenning.

File: procrustes/woodgate.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    ----------
+    x : np.ndarray
+        The matrix to be checked.
+
+    Returns
+    -------
+    bool
+        True if the matrix is positive semidefinite.
+    """
+    return np.all(np.linalg.eigvals(x) >= 0)
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: That is true, but won't it hamper readability within the `woodgate` function where this is method is being called.

File: procrustes/woodgate.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+    Returns
+    -------
+    np.ndarray
+        The re-constructed matrix.
+    """
+    eigenvalues, U = np.linalg.eig(arr)
+    eigenvalues_pos = [max(0, i) for i in eigenvalues]
+    U_inv = np.linalg.inv(U)
+    return U @ np.diag(eigenvalues_pos) @ U_inv
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: I will fix it in the next commit.

**Comment by banrovegrie** [comment]
Hey, so I ran into an issue while using `ProcrustesResult` to return the error and required transformation. In case of `ProcrustesResult`, we aim to return $S, T$. They are defined as follows.

$$
\|\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\|_{F}^2 =
\text{Tr}\left[
     \left(\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\right)^\dagger
     \left(\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\right)\right]
$$

However, in our program, we are dealing with transformation only on the left-hand side. Namely, $\min_{\mathbf{P}} \|F - \mathbf{P}G\|^2$.

How should I proceed for this?

**Comment by banrovegrie** [comment]
Also, abiding by the terminology present in other modules, $A$ is considered as the matrix on which the transformation will be applied and $B$ is the target matrix.

Thus, our result in `woodgate` is effectively finding a left hand transformation $P$ such that $\|PA - B\|^2 = \|B - PA\|^2$ is minimum. Note that to draw parallel with the paper, $A$ is $G$ and $B$ is the matrix $F$.

**Comment by banrovegrie** [comment]
Other than these two persisting questions, everything else is complete. @FanwangM

File: procrustes/woodgate.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (FanwangM) [reply procrustes/woodgate.py]: If you want to keep it, we need at least to make it a private function as it's only used for once and only used in this module. @banrovegrie

**Comment by FanwangM** [comment]
> Hey, so I ran into an issue while using `ProcrustesResult` to return the error and required transformation. In case of `ProcrustesResult`, we aim to return S,T. They are defined as follows.
> 
> |SAT−B|F2=Tr[(SAT−B)†(SAT−B)]
> 
> However, in our program, we are dealing with transformation only on the left-hand side. Namely, minP|F−PG|2.
> 
> How should I proceed for this?

The other matrix is defined as the identity matrix with the same shape.

**Comment by FanwangM** [comment]
The linter keeps complaining and needs to be fixed.
Also, can you also add a test file for this module, either in this PR or another PR? Thank you.
@banrovegrie

**Comment by banrovegrie** [comment]
Apparently, `tox` says the error is in `softassign.py`. That shouldn't be the case right?

```
************* Module procrustes.softassign
  procrustes/softassign.py:221:15: E1121: Too many positional arguments for method call (too-many-function-args)
```

File: procrustes/woodgate.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: Ok, I will make it private.

**Comment by FanwangM** [comment]
Some general comments for this implementation.

- [ ] Rename `woodgate.py` to `psdp.py`
- [ ] Rename function `woodgate` to `psdp_woodgate` as we have two more algorithms coming.
- [ ] Only put  `psdp_woodgate` as a public function that a user can access and put other functions as private functions
- [ ] Fix the matrix representation in the docstring as we use bold characters to denote matrix,  such as $\mathbf{P}$.
- [ ] Try to avoid use _@_ operations and use `np.dot` if possible, https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html. This can help make things clean and consistent
- [ ] We will need to make the argument names consistent and you can take a look at other examples in Procrustes package for reference.

Thank you for the nice work. @banrovegrie

**Comment by banrovegrie** [comment]
[x] Renamed `woodgate.py` to `psdp.py`.
[x] Renamed function `woodgate` to `psdp_woodgate`.
[x] `psdp_woodgate` is now the only public function.
[x] Bold characters represent matrices in docstring now.

- Testing file I will add in a different PR so that there we can talk about the efficiency of the implementation as well.
- Since, we will use the same testing for multiple algorithms, I think this makes more sense.
- In the case of small equations, I tried to use `np.dot` rather than `@` but some of the formulae are quite elaborate and converting them too to `np.dot` might make the readability hard.

File: procrustes/woodgate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/woodgate.py]: Fixed in https://github.com/theochem/procrustes/pull/170/commits/cee97fa4cc47bf017b24ed181f0f1d3d3e15eb24

**Comment by banrovegrie** [comment]
I tested with `tox` locally. All the tests seem to pass. @FanwangM

**Review by FanwangM — CHANGES_REQUESTED** [review]
The codes seem good for now. A few minor comments about single underscore and double underscore and we should use single underscore for all the private functions.

And we should move `psdp_woodgate` function to the top of all the private functions.
Thanks for the nice work.

File: procrustes/psdp.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+import numpy as np
+from procrustes.utils import ProcrustesResult
+import scipy.linalg as lin
+from scipy.optimize import minimize
+
+
+__all__ = ["psdp_woodgate"]
+
+
+def __permutation_matrix(arr: np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py:37]: Yea true, since we don't have classes just the conventional single `_` will work.

**Comment by banrovegrie** [comment]
Oh okay you wanted me to implement `np.linalg.multi_dot` makes sense. Will do it right away.

File: procrustes/psdp.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py]: fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

File: procrustes/psdp.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+            j = k
+            k += 1
+            p[i, j] = 1
+        else:
+            j += n
+            p[i, j] = 1
+    return p
+
+
+def __no_update(error: List[int], threshold: int = 1e-5) -> bool:
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py]: fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

File: procrustes/psdp.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+
+    Returns
+    -------
+    bool
+        True if the error has reduced/changed.
+    """
+    return abs(error[-1] - error[-2]) < threshold
+
+
+def __is_pos_semi_def(arr: np.ndarray) -> bool:
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py]: fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

File: procrustes/psdp.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+    threshold : int, optional
+        The threshold below which we change isn't
+        considered.
+
+    Returns
+    -------
+    bool
+        True if the error has reduced/changed.
+    """
+    return abs(error[-1] - error[-2]) < threshold
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py]: fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

File: procrustes/psdp.py [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+    ----------
+    Refer to equations (26) and (27) in [1] for exact deinitions of the
+    several terms mentioned in this function.
+    .. [1] Woodgate, K. G. (1996). "A new algorithm for positive
+        semidefinite procrustes". Journal of the American Statistical
+        Association, 93(453), 584-588.
+    """
+    n = e.shape[0]
+    s = np.linalg.matrix_rank(e)
+    v = lin.null_space(le.T).flatten()
```
[the remark this answers is not in the document]
Comment (banrovegrie) [reply procrustes/psdp.py]: fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

**Comment by banrovegrie** [comment]
I am just left to implement `multi_dot`. Once that is done, could you merge it so that I can start pushing the test using another PR and simultaneously start working on the new algorithm?

**Comment by banrovegrie** [comment]
In https://github.com/theochem/procrustes/pull/170/commits/2baf69adc757f7f8163c473c360655d3d12b0555, I have reimplemented `@` with `np.dot` or `np.linalg.mult_dot` as required.

**Comment by banrovegrie** [comment]
@FanwangM could you run tests whenever you are free?

**Review by FanwangM — APPROVED** [review]

**Pull request merged**